lib/private/Preview/Generator.php

```diff
@@ -204,6 +206,95 @@ public function generatePreviews(File $file, array $specifications, $mimeType =
 		return $preview;
 	}
 
+	/**
+	 * @param int $semId
+	 * @param int $concurrency
+	 * @return bool|resource the semaphore on success or false on failure
+	 */
+	public static function guardWithSemaphore(int $semId, int $concurrency) {
+		if (extension_loaded('sysvsem')) {
+			$sem = sem_get($semId, $concurrency);
+			if ($sem === false) {
+				return false;
+			}
+			if (!sem_acquire($sem)) {
+				return false;
+			}
+			return $sem;
+		}
+		return false;
+	}
+
+	/**
+	 * @param resource $sem
+	 * @return bool
+	 */
+	public static function unguardWithSemaphore($sem) {
+		if (!is_resource($sem)) {
+			return false;
+		}
+		if (extension_loaded('sysvsem')) {
+			return sem_release($sem);
+		}
+		return false;
+	}
+
+	/**
+	 * Get the number of concurrent threads supported by the host
+	 *
+	 * @return int number of concurrent threads, or 0 if it cannot be determined
+	 */
+	public static function getHardwareConcurrency() {
+		static $width;
+		if (!isset($width)) {
+			if (is_file("/proc/cpuinfo")) {
+				$width = substr_count(file_get_contents("/proc/cpuinfo"), "processor");
+			} else {
+				$width = 0;
+			}
+		}
+		return $width;
+	}
+
+	/**
+	 * Get number of concurrent preview generations from system config
+	 *
+	 * Two config entries, `preview_concurrency_new` and `preview_concurrency_all`,
+	 * are available. If not set, the default values are determined with the hardware concurrency
+	 * of the host. In case the hardware concurrency cannot be determined, or the user sets an
+	 * invalid value, fallback values are:
+	 * For new images whose previews do not exist and need to be generated, 4;
+	 * For all preview generation requests, 8.
+	 * Value of `preview_concurrency_all` should be greater than or equal to that of
+	 * `preview_concurrency_new`, otherwise, the latter is returned.
+	 *
+	 * @param string $type either `preview_concurrency_new` or `preview_concurrency_all`
+	 * @return int number of concurrent preview generations, or -1 if $type is invalid
+	 */
+	public function getNumConcurrentPreviews($type) {
+		static $cached = array();
+		if (array_key_exists($type, $cached)) {
+			return $cached[$type];
+		}
+
+		$hardwareConcurrency = self::getHardwareConcurrency();
+		switch ($type) {
+			case "preview_concurrency_all":
+				$fallback = $hardwareConcurrency > 0 ? $hardwareConcurrency * 2 : 8;
```

Comment:
```suggestion
				$fallback = $hardwareConcurrency > 0 ? $hardwareConcurrency : 4;
```
I've tested this a bit and I have tried setting preview_concurrency_all to the same like preview_concurrency_new and actually this feels better if both values are equal since it doesn't wait until the next two batches previews are generated but displays them directly after one batch.

Reply:
Maybe it makes sense to merge preview_concurrency_all and preview_concurrency_new into one?
(as I found out, it makes sense to have them equal)

Reply:
I think it's probably better to leave `preview_concurrency_all` greater than `preview_concurrency_new`. The logic of preview generation is: 
1. check if the "max preview" of a requested image has been generated and cached. Concurrent execution is limited by the `preview_concurrency_all` semaphore.
2. if so, we can simply grab it and return. This is the light path, the `preview_concurrency_new` semaphore is not involved at all;
3. otherwise, it will be generated. This is the heavy path that uses extensive CPU and memory resources, and concurrent execution is limited by the `preview_concurrency_new` semaphore;

The `preview_concurrency_all` exists for the light path. Consider a host environment with 4 cores, and `preview_concurrency_new` is set to 2. When a photo album being browsed contains both old pics that has preview generated already, as well as new pics, it could be that some requests go through the heavy path while some go through the light path. We want to make sure those that go through the light path can be processed as quickly as possible, so setting `preview_concurrency_all` to the same value as `preview_concurrency_new` will block them, because there is no capacity left. Setting `preview_concurrency_all` to 4 will make sure while 2 are busy processing new pics, the other 2 cores are free to process those light requests.  Therefore a larger `preview_concurrency_all` makes sense.

Reply:
Okay, thanks for the clarification! 
Makes sense -> closing this review.